src/fparser/Fortran2003.py

```diff
@@ -4991,184 +5002,327 @@ class End_Select_Type_Stmt(EndStmtBase): # R824
+    R830
+
+    <loop-control> = [ , ] <do-variable> = scalar-int-expr,
+                                           scalar-int-expr
+                                           [ , <scalar-int-expr> ]
                      | [ , ] WHILE ( <scalar-logical-expr> )
     """
     subclass_names = []
     use_names = ['Do_Variable', 'Scalar_Int_Expr', 'Scalar_Logical_Expr']
+
+    @staticmethod
     def match(string):
+        '''
+        :param str string: Fortran code to check for a match
+        :return: 3-tuple containing strings and instances of the classes
+                 determining loop control (optional comma delimiter,
+                 optional scalar logical expression describing "WHILE"
+                 condition or optional counter expression containing loop
+                 counter and scalar integer expression)
+        :rtype: 3-tuple of objects or nothing for an "infinite loop"
+        '''
+        # pylint: disable=unbalanced-tuple-unpacking
+        optional_delim = None
+        # Match optional delimiter
         if string.startswith(','):
             line, repmap = string_replace_map(string[1:].lstrip())
+            optional_delim = ", "
         else:
             line, repmap = string_replace_map(string)
-        if line[:5].upper()=='WHILE' and line[5:].lstrip().startswith('('):
-            l = line[5:].lstrip()
-            i = l.find(')')
-            if i!=-1 and i==len(l)-1:
-                return Scalar_Logical_Expr(repmap(l[1:i].strip())),
-        if line.count('=')!=1: return
-        var,rhs = line.split('=')
+        # Match "WHILE" scalar logical expression
+        if line[:5].upper() == 'WHILE' and line[5:].lstrip().startswith('('):
+            lbrak = line[5:].lstrip()
+            i = lbrak.find(')')
+            if i != -1 and i == len(lbrak)-1:
+                scalar_logical_expr = \
+                    Scalar_Logical_Expr(repmap(lbrak[1:i].strip()))
+                return scalar_logical_expr, None, optional_delim
+        # Match counter expression
+        if line.count('=') != 1:
+            return
+        var, rhs = line.split('=')
         rhs = [s.strip() for s in rhs.lstrip().split(',')]
-        if not 2<=len(rhs)<=3: return
-        return Variable(repmap(var.rstrip())), \
-                list(map(Scalar_Int_Expr, list(map(repmap,rhs))))
-    match = staticmethod(match)
+        if not 2 <= len(rhs) <= 3:
+            return
+        counter_expr = (Variable(repmap(var.rstrip())),
+                        list(map(Scalar_Int_Expr, list(map(repmap, rhs)))))
+        return None, counter_expr, optional_delim
+
     def tostr(self):
-        if len(self.items)==1: return ', WHILE (%s)' % (self.items[0])
-        return ', %s = %s' % (self.items[0], ', '.join(map(str,self.items[1])))
+        '''
+        :return: parsed representation of loop control construct
+        :rtype: string
+        '''
+        # pylint: disable=unbalanced-tuple-unpacking
```

Comment:
I think pylint is confused because it cannot recognise that the number of items in self.items returned from match(string) method is the same as the number of values unpacked from it in the tostr method.

E:7603, 8: Possible unbalanced tuple unpacking with sequence: left side has 4 label(s), right side has 0 value(s) (unbalanced-tuple-unpacking)

Possible unbalanced tuple unpacking with sequence%s: left side has %d label(s), right side has %d value(s) Used when there is an unbalanced tuple unpacking in assignment